test/internal/fifo.test.ts

```diff
@@ -0,0 +1,235 @@
+
+  it("has an initial capacity of 4", () => {
+    expect(queue._list.length).toBe(4);
+    expect(queue._capacityMask).toBe(3);
+  });
+
+  it("toArray() returns an empty array", () => {
+    expect(queue.toArray()).toEqual([]);
+  });
+
+  describe("When an element is pushed", () => {
+    beforeEach(() => {
+      queue.push(42);
+    });
+
+    it("has a size of 1", () => {
+      expect(queue.size()).toBe(1);
+    });
+
+    it("can be peeked without removing it", () => {
+      expect(queue.peek()).toBe(42);
+      expect(queue.size()).toBe(1);
+    });
+
+    it("is not empty", () => {
+      expect(queue.isEmpty()).toBe(false);
+      expect(queue.isNotEmpty()).toBe(true);
+    });
+
+    it("can be shifted out", () => {
+      const el = queue.shift();
+      expect(el).toBe(42);
+      expect(queue.size()).toBe(0);
+      expect(queue.isEmpty()).toBe(true);
+    });
+  }); // </When an element is pushed>
+}); // </Given an empty queue>
+
+describe("grow boundary conditions", () => {
+  describe.each([3, 4, 16])("when %d items are pushed", n => {
+    let queue: Dequeue<number>;
+
+    beforeEach(() => {
+      queue = new Dequeue();
+      for (let i = 0; i < n; i++) {
+        queue.push(i);
+      }
+    });
+
+    it(`has a size of ${n}`, () => {
+      expect(queue.size()).toBe(n);
+    });
+
+    it("is not empty", () => {
+      expect(queue.isEmpty()).toBe(false);
+      expect(queue.isNotEmpty()).toBe(true);
+    });
+
+    it(`can shift() ${n} times`, () => {
+      for (let i = 0; i < n; i++) {
+        expect(queue.peek()).toBe(i);
+        expect(queue.shift()).toBe(i);
+      }
+      expect(queue.size()).toBe(0);
+      expect(queue.shift()).toBe(undefined);
+    });
+
+    it("toArray() returns [0..n-1]", () => {
+      // same as repeated push() but only allocates once
+      var expected = new Array<number>(n);
+      for (let i = 0; i < n; i++) {
+        expected[i] = i;
+      }
+      expect(queue.toArray()).toEqual(expected);
+    });
+  });
+}); // </grow boundary conditions>
+
+describe("adding and removing items", () => {
+  let queue: Dequeue<number>;
+  let expected: DequeueList<number>;
+
+  describe("when 10k items are pushed", () => {
+    beforeEach(() => {
+      queue = new Dequeue();
+      expected = new DequeueList();
+
+      for (let i = 0; i < 10_000; i++) {
+        queue.push(i);
+        expected.push(i);
+      }
+    });
+
+    it("has a size of 10000", () => {
+      expect(queue.size()).toBe(10_000);
+      expect(expected.size()).toBe(10_000);
+    });
+
+    describe("when 10 items are shifted", () => {
+      beforeEach(() => {
+        for (let i = 0; i < 10; i++) {
+          expect(queue.shift()).toBe(expected.shift());
+        }
+      });
+
+      it("has a size of 9990", () => {
+        expect(queue.size()).toBe(9990);
+        expect(expected.size()).toBe(9990);
+      });
+    });
+
+    describe("when 1k items are pushed, then removed", () => {
+      beforeEach(() => {
```

Comment:
i don't think the test itself should exist in a beforeEach?

Reply:
Woops fixed haha